diff --git a/contrib/PostInstall.cmake.in b/contrib/PostInstall.cmake.in
index f6b9a27..50e600f 100644
--- a/contrib/PostInstall.cmake.in
+++ b/contrib/PostInstall.cmake.in
@@ -3,18 +3,18 @@ message(STATUS "Performing post install steps")
 set (PYTHON_EXECUTABLE @PYTHON_EXECUTABLE@)
 
 # cleanup hard-coded links to libraries
-if (WIN32)
+if (1)
 
     execute_process(
-      COMMAND ${PYTHON_EXECUTABLE} @CMAKE_SOURCE_DIR@/contrib/remove-library.py  ${CMAKE_INSTALL_PREFIX}/@CONFIG_INSTALL_DIR@/tixi-targets-release.cmake "@CURL_LIBRARIES@;"
+      COMMAND ${PYTHON_EXECUTABLE} @CMAKE_SOURCE_DIR@/contrib/remove-library.py  ${CMAKE_INSTALL_PREFIX}/@CONFIG_INSTALL_DIR@/tixi-targets-release.cmake "@CURL_LIBRARIES@"
     )
 
     execute_process(
-      COMMAND ${PYTHON_EXECUTABLE} @CMAKE_SOURCE_DIR@/contrib/remove-library.py  ${CMAKE_INSTALL_PREFIX}/@CONFIG_INSTALL_DIR@/tixi-targets-release.cmake "@XSLT_LIBRARIES@;"
+      COMMAND ${PYTHON_EXECUTABLE} @CMAKE_SOURCE_DIR@/contrib/remove-library.py  ${CMAKE_INSTALL_PREFIX}/@CONFIG_INSTALL_DIR@/tixi-targets-release.cmake "@XSLT_LIBRARIES@"
     )
 
     execute_process(
-      COMMAND ${PYTHON_EXECUTABLE} @CMAKE_SOURCE_DIR@/contrib/remove-library.py  ${CMAKE_INSTALL_PREFIX}/@CONFIG_INSTALL_DIR@/tixi-targets-release.cmake "@XML_LIBRARIES@;"
+      COMMAND ${PYTHON_EXECUTABLE} @CMAKE_SOURCE_DIR@/contrib/remove-library.py  ${CMAKE_INSTALL_PREFIX}/@CONFIG_INSTALL_DIR@/tixi-targets-release.cmake "@XML_LIBRARIES@"
     )
 
 endif()
