diff --git a/bindings/python/make_tixi_python.py b/bindings/python/make_tixi_python.py
index b3b1698..5ecd2e0 100644
--- a/bindings/python/make_tixi_python.py
+++ b/bindings/python/make_tixi_python.py
@@ -123,7 +123,7 @@ def checkAttribute(self, elementPath, attributeName):
     parser.parse_header_file(tixipath + '/src/tixi.h')
     
     # create the wrapper
-    pg = PG.PythonGenerator(name_prefix = 'tixi3', libname = 'tixi3')
+    pg = PG.PythonGenerator(name_prefix = 'tixi', libname = 'tixi3')
     pg.license = apache
     pg.userfunctions = userfunctions
     pg.blacklist = blacklist
