diff --git CMakeLists.txt CMakeLists.txt
index 5a78635..29f4d47 100644
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -9,6 +9,12 @@ if(NOT DEFINED CMAKE_BUILD_TYPE)
     "MinSizeRel" "RelWithDebInfo")
 endif()
 
+set(CMAKE_MACOSX_RPATH ON)
+set(CMAKE_SKIP_BUILD_RPATH FALSE)
+set(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
+set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
+set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+
 project (TIGL)
 set(TIGL_VERSION_MAJOR 2)
 set(TIGL_VERSION_MINOR 1)
diff --git bindings/python_internal/CMakeLists.txt bindings/python_internal/CMakeLists.txt
index cfa9b81..c4d22c0 100644
--- bindings/python_internal/CMakeLists.txt
+++ bindings/python_internal/CMakeLists.txt
@@ -75,7 +75,7 @@ if (TIGL_PYTHON_INTERNAL)
         swig_add_module(${MODULE} python ${MODULE}.i )
         swig_link_libraries(${MODULE} ${PYTHON_LIBRARIES} TIGL)
 
-        install(FILES ${CMAKE_CURRENT_BINARY_DIR}/_${MODULE}.${EXTENSION}
+        install(TARGETS _${MODULE}
                 DESTINATION share/tigl/python/internal
                 COMPONENT interfaces
         )
